### PR TITLE
feat: dynamic value implementation

### PR DESCRIPTION
**Pull request by dbouwman**
1. Description:

Adds types and functions for dynamic value definition and resolution. This is the underlying subsystem that will power Metrics as well as run-time interpolation of layouts, and deploy time interpolation of templates

1. Instructions for testing: run tests

1. Closes Issues: #[6326](https://devtopia.esri.com/dc/hub/issues/6326)

1. [x] used semantic commit messages
  
1. [x] PR title follows semantic commit format (**CRITICAL** if the title is not in a semantic format, the release automation will not run!)

1. [x] updated `peerDependencies` as needed. **CRITICAL** our automated release system can **not** be counted on to update `peerDependencies` so we _must_ do it manually in our PRs when needed. See the [updating peerDependencies](/RELEASE.md#Updating-peerDependencies) section of the release instructions for more details.

### CODE REVIEW

File: karma.conf.js
Comment (dbouwman) [inline karma.conf.js:1]: Not sure why the formatting keeps shifting on this file - other than it's clear someone has a different prettier config than I do :/

File: packages/common/src/core/types/DynamicValues.ts [hunk 1]
```diff
@@ -0,0 +1,176 @@
+import { IQuery } from "../../search/types/IHubCatalog";
+import { IConfigurationSchema } from "../schemas";
+
+/**
+ * The source for a dynamic value. Used to determine how to resolve the value
+ */
+export type DynamicValueSource =
```
Comment (tomwayson) [inline packages/common/src/core/types/DynamicValues.ts:7]: should this be `DynamicValueType`, or should we change `IBaseValueDefinition` to have `source: DynamicValueSourece`?

File: packages/common/src/core/types/DynamicValues.ts [hunk 2]
```diff
@@ -0,0 +1,176 @@
...
+  /**
+   * Enables a static value to be provided vs forcing a dynamic value
+   */
+  value: string | number;
+}
+
+/**
+ * Definition for a dynamic value that is resolved from the portal/self
+ */
+export interface IDynamicPortalQueryDefinition extends IBaseValueDefinition {
```
Comment (tomwayson) [inline packages/common/src/core/types/DynamicValues.ts:76]: `IDynamicPortalSelfDefinition`?

File: packages/common/src/utils/internal/resolveDynamicValues.ts [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+export const resolveDynamicValues = async (
+  dynamicValues: DynamicValueDefinition[],
+  context: IArcGISContext
+): Promise<Record<string, DynamicValueResult>> => {
+  let result: Record<string, any> = {};
+
+  for (const valueDef of dynamicValues) {
+    const val = await resolveDynamicValue(valueDef, context);
+    result = { ...result, ...val };
+  }
```
Comment (tomwayson) [inline packages/common/src/utils/internal/resolveDynamicValues.ts:25]: i assume you have a reason not to use `Promise.all()` - if so maybe document w/ a comment?

**Review by tomwayson — APPROVED** [review]
LGTM minor thoughts on naming

File: packages/common/src/utils/internal/resolveDynamicValues.ts (hunk 3 above)
Comment (dbouwman) [reply packages/common/src/utils/internal/resolveDynamicValues.ts:25]: tbh - I don't have a solid reason for not parallelizing this... just did not want to mess with it b/c it works :) 
I'll re-assess and either use Promise.all or leave a comment

**Pull request merged**